rero_ils/modules/selfcheck/api.py

```diff
@@ -427,3 +429,70 @@ def selfcheck_checkin(transaction_user_pid, item_barcode, **kwargs):
                 _('Error encountered: please contact a librarian'))
             raise SelfcheckCirculationError('self checkin failed', checkin)
         return checkin
+
+
+def selfcheck_renew(transaction_user_pid, item_barcode, **kwargs):
+    """SIP2 handler to perform renew.
+
+    perform renew action received from the selfcheck.
+    """
+    if check_sip2_module():
+        from invenio_sip2.errors import SelfcheckCirculationError
+        from invenio_sip2.models import SelfcheckFeeType, SelfcheckRenew
+
+        terminal = SelfcheckTerminal.find_terminal(
+            name=kwargs.get('terminal'))
+        item = Item.get_item_by_barcode(
+            barcode=item_barcode,
+            organisation_pid=terminal.organisation_pid
+        )
+        document = Document.get_record_by_pid(item.document_pid)
+
+        renew = SelfcheckRenew(
+            title_id=title_format_text_head(document.get('title'))
+        )
+        with current_app.test_request_context() as ctx:
+            language = kwargs.get('language', current_app.config
+                                  .get('BABEL_DEFAULT_LANGUAGE'))
+            ctx.babel_locale = language
+            try:
+                staffer = Patron.get_record_by_pid(transaction_user_pid)
+                if staffer.is_librarian:
+                    # get renewal count
+                    renewal_count = item.get_extension_count()
+                    if renewal_count > 1:
+                        renew['renewal'] = True
+                    # do extend loan
+                    result, data = item.extend_loan(
+                        transaction_user_pid=staffer.pid,
+                        transaction_library_pid=terminal.library_pid,
+                        item_pid=item.pid,
+                        selfcheck_terminal_id=str(terminal.id),
+                    )
+                    if data[LoanAction.EXTEND]:
+                        loan_pid = data[LoanAction.EXTEND].get('pid')
+                        loan = Loan.get_record_by_pid(loan_pid)
```

Comment:
Sure that you need to get the loan again? Is it not returned by the `data[LoanAction.EXTEND]`?

Reply:
you are right. thanks